src/engine/thread.js

```diff
@@ -388,8 +409,8 @@ class Thread {
      */
     isRecursiveCall (procedureCode) {
         let callCount = 5; // Max number of enclosing procedure calls to examine.
-        const sp = this.stack.length - 1;
-        for (let i = sp - 1; i >= 0; i--) {
+        const sp = this.stack.length;
+        for (let i = sp - 1; i >= 1; i--) {
```

Comment:
Why don't we look at `stack[0]`?

Reply:
The operations to the stack in this PR leave it with the same number of stack members by making the first member null. So its not a real stack value, but gives this PR the length side effect, and restores pointer to null on the last popStack.

Basically its done this way to fulfill existing tests that look at stack length.

Reply:
Hmmm that seems a little too subtle/scary from a maintenance perspective. We should probably remove the null from the beginning of the stack and update the tests...

Reply:
Sounds good to me. That'll also help with tests that explicitly check for popStack returning undefined.

Reply:
Removing the null from the stack turned out kind of weird. Primarily code that tested `stack.length === 0` could not switch to `stack.pointer === null` but instead `thread.stackFrame === null`. The pointer/stack data can be null to mean the end of a stack and not just the end of the thread. The pointer-less/all-stack version uses `.length` of 1 or more while `stack[0] === null`. Representing its in a stack but at the end.

So this makes `pointer === null` a little ambiguous but really points to using `stackFrame === null` instead. Since it is only null when the thread is not working.

Reply:
Ah, okay. Where did we end up landing with this?

Reply:
I think we landed on `stack[0]` cannot be null. Or at least that is currently the implementation in the PR. In this PR currently, values in stack will always be a blockId. `pointer` can be null, but none of the values in stack will be null.